### PR TITLE
Issue 363: Nodejs publish

### LINKED ISSUE #363: Nodejs publish
**Problem description**
Build and publish the Nodejs bindings for different systems and architectures.

**Problem location**
`./nodejs/`

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by thekingofcity**
**Change log description**  
Build and publish the Nodejs bindings for different systems and architectures.

**Purpose of the change**  
Fix #363 

**What the code does**  
It basically pretties the docs and updates CI so npm packages are published to the npm registry and platform-specific binaries are released at Github releases pages.
For more details about the release process, see https://github.com/neon-bindings/neon/issues/117#issuecomment-1073561462

This should be merged after #360.

**How to verify it**  
`tagPublish` workflow will publish npm packages and platform-specific binaries are released at Github releases pages.
See my own fork's actions and release page for the expected outcome.

https://www.npmjs.com/package/@thekingofcity/pravega
https://github.com/thekingofcity/pravega-client-rust/actions/runs/2014604184
https://github.com/thekingofcity/pravega-client-rust/releases/tag/v0.4.12

Release of this tag must be created so binaries can be uploaded!

### CODE REVIEW

File: nodejs/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+authors = ["Pravega Community"]
+categories = ["network-programming"]
+description = "Pravega client"
+edition = "2018"
+exclude = ["pravega.node"]
+keywords = ["streaming", "client", "pravega"]
+license = "Apache-2.0"
+name = "pravega_nodejs"
+readme = "README.md"
+repository = "https://github.com/pravega/pravega-client-rust"
```
Comment (tkaitchuck) [inline nodejs/Cargo.toml:11]: Should have a description

File: nodejs/README.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+This project provides a way to interact with [Pravega](https://cncf.pravega.io) via a Nodejs client.
+
+Pravega is an open source distributed storage service implementing Streams. It offers Stream as the main primitive for the foundation of reliable storage systems: a high-performance, durable, elastic, and unlimited append-only byte stream with strict ordering and consistency.
+
+## Install
+
+The client library can be installed using npm or yarn.
+
+```shell
+npm install @thekingofcity/pravega
```
Comment (tkaitchuck) [inline nodejs/README.md:12]: This should be generic

File: nodejs/README.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    const seg_slice = await stream_reader.get_segment_slice();
+    const dec = new TextDecoder('utf-8');
+    for (const event of seg_slice) {
+        const raw_bytes = event.data();
+        console.log(`Event at ${event.offset()} reads ${dec.decode(raw_bytes)}`);
+    }
+    // Release the current slice so other reader can lock and read this slice.
+    stream_reader.release_segment(seg_slice);
+} catch (e) {
+    console.log(e);
```
Comment (tkaitchuck) [inline nodejs/README.md:79]: Do we need the catch?

File: nodejs/package.json [hunk 4]
```diff
@@ -0,0 +1,73 @@
+{
+  "name": "@thekingofcity/pravega",
```
Comment (tkaitchuck) [inline nodejs/package.json:2]: This should be generic

File: nodejs/package.json [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      }
+    },
+    "moduleNameMapper": {
+      "^(\\.{1,2}/.*)\\.js$": "$1"
+    }
+  },
+  "binary": {
+    "module_name": "pravega",
+    "module_path": "./native",
+    "host": "https://github.com/thekingofcity/pravega-client-rust/releases/download/",
```
Comment (tkaitchuck) [inline nodejs/package.json:69]: Same here

File: nodejs/README.md (hunk 3 above)
Comment (thekingofcity) [reply nodejs/README.md]: Removed.

File: nodejs/README.md (hunk 2 above)
Comment (thekingofcity) [reply nodejs/README.md:12]: I want to show that the current combination would result in a working package at my namespace. But turning it to something like `@pravega/pravega` requires far more effort on both npm registry and github actions. Things like using which email to register the npm and whether the binary url will work for the new address.

**Comment by thekingofcity** [comment]
Distribution of commonjs should be supported as there are still legacy environment that deeply rely it (electron for example). Below are introduction about `cjs`/`esm` and best practice to support both js modules.

> https://dev.to/iggredible/what-the-heck-are-cjs-amd-umd-and-esm-ikm
> https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html

**Comment by thekingofcity** [comment]
There should be an `NPM_TOKEN` secret before release.

**Comment by tkaitchuck** [comment]
Is this ready for review?

**Comment by thekingofcity** [comment]
> Is this ready for review?

Of course, I just want to make sure this PR is not merged before the secrets and npm account are added.

We need to at least create an npm account with an email that the pravega team has access to and create a package named `@pravega/pravega`.

**Comment by thekingofcity** [comment]
This is ready to be merged if `NPM_TOKEN` secret is set.

**Review by tkaitchuck — APPROVED** [review]

**Pull request merged**